### ISSUE TITLE
Functional Rates: token_number expression (#(ID)) excludes zero

### ISSUE THREAD

**Issue by paternostrox** [sarahtattersall/PIPE#119]
I really got the impression that by using token_number expression (#(ID)), the output should be the real number of tokens in a place, so using purely "#(ID)" in an arc, the arc should check how many tokens are in ID, and consume that amount, no matter what is the number... But for some reason that doesn't work with zero, is this expected to happen?

Image of  a test, as you can see T0 is not enabled and will only be enabled if P0 has one or more tokens.

![asda](https://user-images.githubusercontent.com/19597048/28448064-7c0a0702-6dab-11e7-99b5-383f0e4bfdc7.png)

**Comment by sjdayday** [comment]
Yes, this is working as intended.  In the simplest Petri net definition, tokens are integer-valued, and a place may have 0 or more tokens.  An arc weight determines the consumption & production of tokens; the smallest number of tokens that can be consumed or produced is one.

Hope that makes things more clear; please let me know if you have other questions.

best,
Steve

**Comment by paternostrox** [comment]
Thank you for the quick reply. I really thought that for example, setting weight zero on arc P0 to T0 (previous image), meant the arc condition is fulfilled when P0 is empty, and T0 should be enabled then. I'm sure that happened before, but now testing it again I guess it was a bug.

That feature would really help me to get a simpler reset, I tried to reduce what I was trying to do so I can show you (all the transitions priorities are the same):

![reset](https://user-images.githubusercontent.com/19597048/28490580-13fb283a-6eb4-11e7-80dc-04eba9b7c881.png)

I have this situation where the number of tokens I have to remove is unsure, but I just need to remove them all no matter what the number is.. I have this in a larger scale in my actual project and I'm trying to find a solution that doesn't lead to organizational mayhem. Any thoughts?

Thanks in advance.

**Comment by sjdayday** [comment]
I think what you're doing should work.  If you just want to empty all tokens from a place, just have an inbound arc with weight #(Pn) and no outbound arc.

Steve

**Comment by paternostrox** [comment]
It really doesn't, let me animate the current petrinet. Note that in my net I want to be able to get into lockdown, this is expected:
![lockdown](https://user-images.githubusercontent.com/19597048/28493403-c50a4f06-6eec-11e7-94a5-9377e133fcae.png)

But we also get this unexpected lockdown, here I want to be able to reset, but I can't think on any solution?
![resetlockdownunexpected](https://user-images.githubusercontent.com/19597048/28493404-c96a0fbe-6eec-11e7-9b95-12954a73b4d6.png)

Current example XML:
[reset.zip](https://github.com/sarahtattersall/PIPE/files/1167527/reset.zip)

**Comment by sjdayday** [comment]
Can you explain more about what you are trying to model?  What are the transitions and places attempting to express?

**Comment by sjdayday** [comment]
If T0 fires, you could use an inhibitory arc from P3 to T3 to keep T3 from firing and ending up with a second token.  This would keep tokens in P1, P3 and ShouldReset, so that Reset would be enabled, but T2 would also be enabled, so no guarantee that Reset would fire.  

Bottom line:  any firing pattern can be programmed, with enough work.  You need to decide what you want to model to drive the appropriate Petri net structure.

Steve
